### DISCUSSION TITLE
Best pattern for caching millions of pages (app directory)

### DISCUSSION THREAD

**Discussion by lzrscg** [vercel/next.js#49951] [Help]
### Summary

Hi all! I hope this is the right place to ask this question. For context, we are using the **app directory** and Vercel for hosting.

We have many millions of pages, but would like to take advantage of caching and static generation as much as possible.

My initial approach would be to statically generate params of the _most important pages_, and then set `dynamicParams` to `true`. So at least some pages will get the benefit of static generation, while the rest can be dynamically rendered.

However, in our case, the most important pages are usually the most recent pages. So, as new content is added to our site, it would be ideal to generate pages in real-time instead of going through a list at build time. Is there any mechanism that can be used to statically generate or cache page **after** build time? It's not clear to me how ISR works with the app dir. Ideally, we could do something similar to ISR, in which we can pre-render pages once they have been created (however, we don't need to revalidate them in the future)

I am also interested in hearing other ideas dealing with large numbers of pages.

Thank you for reading my question and I would much appreciate any advice.

P.S. this may be more of a Vercel question than Next question. But what are the current the limits on statically generated pages? Is it just based on the build time (45min) and disk space limit (13GB)?

**Comment by devjiwonchoi** [comment]
I would suggest to separate API endpoints as "recents" API and "overall" API. 
"Recents" API queries from DB with a limited amount of numbers and order by latest id. 
By doing this, you can call "recents" API by getStaticParams, and generate the routes at build time. 
For the "overall" API, you can set them as dynamicParams, so that they would be rendered on-demand.

For the answer to the last question, if you use dynamic routes, Next.js does not generate HTML but only routes by default.
On the other hand, if you have numerous static pages rendered as SSR, then yes it would depend on your disk space.

Feel free to give feedback 😄

**Reply by lzrscg** [reply]
Thanks for the response! I can definitely iterate through the top/recent pages to statically generate them using `generateStaticParams`.

I am curious if there is a way to trigger this **after** build time, though? I want to be able to generate a new pages as they come online.

Or perhaps there is some similar caching technique that can pre-fetch API calls before the page renders for users (to be clear, I'm not talking about pre-fetching on the client side)? Then, although the page would not be statically generated, at least it would be rendered dynamically as quickly as possible.

**Reply by devjiwonchoi** [reply]
As far as I know, you cannot do it **after buildtime**.

In my opinion, your option seems to be the best at this point.

Since generateStaticParams does not build HTML but only routes, if you use Dynamic Routes you have to bite that
HTML will be rendered on-demand.

> [Dynamic Rendering](https://nextjs.org/docs/app/building-your-application/rendering#dynamic-rendering)
With Dynamic Rendering, both Server and Client Components are rendered on the server at request time. The result of the work is not cached.
